File: solutions/problem_1415_the_k_th_lexicographical_string_of_all_happy_strings_of_length_n/solution.py
"""
Problem 1415: The k-th Lexicographical String of All Happy Strings of Length n
============================================================================
Difficulty: Medium
Tags: String, Backtracking
Pattern: Backtracking

Time Complexity:  O(k^n) or O(n!)
Space Complexity: O(n)
"""

class Solution:
    def getHappyString(self, n: int, k: int) -> str:
        # Backtracking - O(2^n) or O(n!) time
        result = []
        
        def backtrack(path, start):
            result.append(path[:])
            for i in range(start, len(n)):
                path.append(n[i])
                backtrack(path, i + 1)
                path.pop()
        
        backtrack([], 0)
        return result

File: solutions/problem_1415_the_k_th_lexicographical_string_of_all_happy_strings_of_length_n/solution.cpp
/*
 * Problem 1415: The k-th Lexicographical String of All Happy Strings of Length n
 * ============================================================================
 * Difficulty: Medium
 * Tags: String, Backtracking
 * Pattern: Backtracking
 *
 * Time Complexity:  O(k^n) or O(n!)
 * Space Complexity: O(n)
 */

#include <functional>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    string getHappyString(int n, int k) {
        // Backtracking - O(2^n) or O(n!) time
        vector<vector<int>> result;
        vector<int> path;
        function<void(int)> backtrack = [&](int start) {
            result.push_back(path);
            for (int i = start; i < (int)n.size(); i++) {
                path.push_back(n[i]);
                backtrack(i + 1);
                path.pop_back();
            }
        };
        backtrack(0);
        return result;
    }
};
